rfcs/20190821-nodejs-saved-model.md

```diff
@@ -0,0 +1,401 @@
+
+This could be a solution for stage 1 of this project. Getting input/output names through tfjs-node API can be a stage 2 of this project and implement parallelly with other part of this project. Thus in long term users do not need to rely on TF Python or other tools to execute a SavedModel in tfjs-node.
+
+Pros:
+
+*   Getting input/output names will not block other part of this project
+*   Users could start to execute SavedModel in Node.js sooner
+
+Cons:
+
+*   Users need to use saved_model_cli to inspect the SavedModel
+*   Still need to figure out a solution to get default input/output name in long term so users do not need to know the details of the SavedModel
+
+
+##### Option 2: add new TF C API to get input/output names from meta graph definition
+
+Tfjs-node is running the TensorFlow C API, which provides TF C API header files and shared libraries. It does not require any Protocol Buffers to call the library. So adding new TF API to get serving_default input/output name string from a SavedModel would enable projects like tfjs-node to execute a session without bringing Protocol Buffers onboard.
+
+Pros:
+
+*   No need to add Protocol Buffers in tfjs-node
+
+Cons:
+
+*   Require review from TF team and extra release cycle for TF C shared libraries
+
+
+##### Option 3: add Protocol Buffers in JavaScript
+
+Like discussed earlier, Protocol Buffers is available in JavaScript and SavedModel proto object can be parsed. So the default signatureDef can be parsed in JavaScript and be used when executing SavedModel session. More details of the SavedModel are available for users to inspect through pure JavaScript API and all the signatureDefs can be executed.
+
+Pros:
+
+*   Users do not need to be familiar with the details of SavedModel or deal with other TF tools
+*   Users could execute any signature definition they want
+
+Cons:
+
+*   Need to compile and maintain SavedModel proto in JS
+
+
+##### Option 4: add Protocol Buffers in tfjs-node c++ addon
+
+Another option similar to option 3 is to add Protocol Buffers in the node native C++ addon. But it requires the tfjs-node package to deliver protocol buffer runtime, which is different for each operating system.
+
+Pros:
+
+*   Users do not need to be familiar with the details of SavedModel or deal with other TF tools
+*   Users could execute any signature definition they want
+
+Cons:
+
+*   Heavy maintenance duty for tfjs-node
+*   Extra tfjs_bindings API to maintain
+*   Need to copy all proto files from TensorFlow to tfjs-node
+
+
+##### Decision
+
+Option 1, ask user to provide the input/output names, could be the temp solution for stage 1 of this project.
+
+Option 2, add function in TF C API to get input/output op names from SavedModel, is a good long term solution. It separates tfjs-node from TF C library details and keeps tfjs_bindings API simple, but it requires extra release cycle for the next TensorFlow C library.
```

Comment:
I would agree that adding the TF C API is a good long term solution that would benefit other language bindings, old and new.  For instance, the current Go and Java implementation for these same functions can be factored out and replaced with the new API calls, simplifying code maintenance.  Should this approach be proposed in parallel while option 1 is used as the temporary solution?

Reply:
Yes that's the plan. I'm updating the doc to list all the changes need to be added in the TF C API, which should benefit all bindings.

Reply:
The latest design is to add protocol buffers in JavaScript, so 1) js developers could inspect the signatures of a SavedModel and 2) no need to add extra N-APIs to make the bindings more complicate and hard to maintain.